Add product tests for access control checks

amrutagokhale opened the pull request:
This PR adds product tests for access control checks that are turned on when `hive.security=sql-standard` is set in `hive.properties` file. The table below shows the list of SQL commands for which Presto adds access control checks (defined in `presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java`) and the privileges required to execute these commands when the above Hive property is turned on. I have written product tests for each of these commands. 

Presto Command | Required Privileges
-------------------- | ---------------------- 
ALTER TABLE	| OWNERSHIP
CREATE TABLE AS | OWNERSHIP OF DATABASE + SELECT ON INPUT TABLE
CREATE VIEW AS | SELECT WITH GRANT OPTION
DELETE | DELETE
DROP TABLE | OWNERSHIP
DROP VIEW | OWNERSHIP
INSERT | 	INSERT
SELECT | SELECT

anusudarsan commented on pom.xml:
```diff
@@ -50,7 +50,7 @@
         <dep.slice.version>0.22</dep.slice.version>
         <dep.aws-sdk.version>1.9.40</dep.aws-sdk.version>
-        <dep.tempto.version>1.8</dep.tempto.version>
+        <dep.tempto.version>1.12</dep.tempto.version>
```
I think presto-product-tests tempto yaml will now need databases.hive.path property

amrutagokhale commented:
@kokosing @anusudarsan can you please review?

Travis failure for one test `testAccessControlCreateView()` is expected. It fails because TD/sprint doesn't have the latest code in prestodb/master. I wrote the test to work with prestodb/master.

anusudarsan commented:
LGTM

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestSqlStandardAccessControlChecks.java:
```diff
@@ -0,0 +1,175 @@
+    @Test(groups = {AUTHORIZATION, HIVE_CONNECTOR, PROFILE_SPECIFIC_TESTS})
+    public void testAccessControlSelect()
+    {
+        String tableName = "alice_owned_table";
```
extract this as contant, there is no need to define this variable in each test

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestSqlStandardAccessControlChecks.java:
```diff
@@ -0,0 +1,175 @@
+    {
+        String tableName = "alice_owned_table";
+        QueryExecutor queryExecutorForAlice = connectToPresto("alice@presto");
+        QueryExecutor queryExecutorForBob = connectToPresto("bob@presto");
```
I would extract creation of the two variables to setup method

```
@BeforeMethodWithContext
public void setUp() {
    this.queryExecutorForAlice = connectToPresto("alice@presto");
    this.queryExecutorForBob = connectToPresto("bob@presto");
}
```

kokosing replied:
moreover I would rename them to `alice` and `bob` respectively. So the code would be much easier to read

```
bob.executeQuery("...");
alice.executeQuery("...");
```

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestSqlStandardAccessControlChecks.java:
```diff
@@ -0,0 +1,175 @@
+        QueryExecutor queryExecutorForAlice = connectToPresto("alice@presto");
+        QueryExecutor queryExecutorForBob = connectToPresto("bob@presto");
+
+        queryExecutorForAlice.executeQuery(format("DROP TABLE IF EXISTS %s", tableName));
```
I think you can move this method to tearDown method

```
@AfterMethodWithContext
public void tearDown() {
  aliice.executeQuer("DROP...");
}
```

kokosing replied:
or even to setUp() method ;)

```
@BeforeMethodWithContext
public void setUp() {
    this.alice = connectToPresto("alice@presto");
    this.bob = connectToPresto("bob@presto");

  alice.executeQuery(format("DROP TABLE IF EXISTS %s", tableName));
  alice.executeQuery(format("CREATE TABLE %s(month bigint, day bigint)", tableName));
}
```

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestSqlStandardAccessControlChecks.java:
```diff
@@ -0,0 +1,175 @@
+
+        queryExecutorForAlice.executeQuery(format("DROP TABLE IF EXISTS %s", tableName));
+        queryExecutorForAlice.executeQuery(format("CREATE TABLE %s(month bigint, day bigint)", tableName));
+        queryExecutorForAlice.executeQuery(format("INSERT INTO %s VALUES (4, 13)", tableName));
```
make assert that data was insserted

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestSqlStandardAccessControlChecks.java:
```diff
@@ -0,0 +1,175 @@
+        assertThat(() -> queryExecutorForBob.executeQuery(format("INSERT INTO %s VALUES (3, 22)", tableName))).
+                failsWithMessage(format("Access Denied: Cannot insert into table default.%s", tableName));
+        queryExecutorForAlice.executeQuery(format("GRANT INSERT ON %s TO bob", tableName));
+        assertThat(queryExecutorForBob.executeQuery(format("INSERT INTO %s VALUES (3, 22)", tableName))).hasRowsCount(1);
```
do select to check that data was inserted

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestSqlStandardAccessControlChecks.java:
```diff
@@ -0,0 +1,175 @@
+
+        assertThat(() -> queryExecutorForBob.executeQuery(format("ALTER TABLE %s ADD COLUMN year bigint", tableName))).
+                failsWithMessage(format("Access Denied: Cannot add a column to table default.%s", tableName));
+        queryExecutorForAlice.executeQuery(format("ALTER TABLE %s ADD COLUMN year bigint", tableName));
```
assert that column was added

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestSqlStandardAccessControlChecks.java:
```diff
@@ -0,0 +1,175 @@
+        assertThat(() -> queryExecutorForBob.executeQuery(format("CREATE VIEW %s AS select * from %s", viewName, tableName))).
+                failsWithMessage(format("Access Denied: Cannot select from table default.%s", tableName));
+        queryExecutorForAlice.executeQuery(format("GRANT SELECT ON %s TO bob", tableName));
+        assertThat(() -> queryExecutorForBob.executeQuery(format("CREATE VIEW %s AS select * from %s", viewName, tableName))).
```
Use `createViewAs()` here and above. Moreover I would extract this view creation sql as variable so reader would not immediately that you are creating the same view each time

amrutagokhale replied:
I extracted the view creation for the two statements above this. It wasn't possible to user `createViewAs` method for them though.

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestSqlStandardAccessControlChecks.java:
```diff
@@ -0,0 +1,175 @@
+        queryExecutorForAlice.executeQuery(format("CREATE VIEW %s AS SELECT * FROM %s", viewName, tableName));
+
+        assertThat(() -> queryExecutorForBob.executeQuery(format("DROP VIEW %s", viewName))).
+                failsWithMessage(format("Access Denied: Cannot drop view default.%s", viewName));
```
no `GRANT DROP VIEW` for bob here?

amrutagokhale replied:
Nope, there is no such `DROP VIEW` privilege that one can grant. Only owners of the view can drop the view.

kokosing commented:
Please ping me on slack once you address comments. Nice product test, however is it possible port this test as unit test?

petroav commented:
@kokosing I don't want to speak for Amruta but my thoughts are that we want to keep this as product test because we want to catch errors when the Hive Metastore semantics/API/internals somehow change such that granting a particular privilege no longer achieves the desired result.

amrutagokhale commented:
@kokosing I have addressed your comments. Please have a look. Also, I am not sure if we can capture these as unit tests. We want to test end-to-end scenarios where a user tries to execute certain operations with and without having sufficient privileges.

kokosing commented:
looks good, thanks

amrutagokhale commented:
Closing this. Prestodb PR is here: https://github.com/prestodb/presto/pull/5953.
